[BoundsSafety] Delay processing of bounds attrs in templates

hnrklssn opened the pull request:
Dynamic bounds attributes do not handle value dependent arguments. To enable wider interop with C++ code bases we delay the processing of these attributes inside templated contexts. Instead we apply the new type while instantiating the function.

One issue with this approach is that instantiation happens after parsing is finished, and the Scope information we use to prevent attributes from referring to arguments from outer scopes is only available during parsing. These diagnostics were emitted at the end of the type processing. In templated contexts we instead perform that analysis during parsing, and delay the rest of the processing. To avoid churn in unrelated tests the rest of the attributes keep their processing as is, until we've investigated what impact hoisting it would have on the user experience.

hnrklssn committed:
[BoundsSafety] Delay processing of bounds attrs in templates

Dynamic bounds attributes do not handle value dependent arguments. To
enable wider interop with C++ code bases we delay the processing of
these attributes inside templated contexts. Instead we apply the new
type while instantiating the function.

One issue with this approach is that instantiation happens after parsing
is finished, and the Scope information we use to prevent attributes from
referring to arguments from outer scopes is only available during
parsing. These diagnostics were emitted at the end of the type
processing. In templated contexts we instead perform that analysis
during parsing, and delay the rest of the processing. To avoid churn in
unrelated tests the rest of the attributes keep their processing as is,
until we've investigated what impact hoisting it would have on the user
experience.

hnrklssn commented:
@swift-ci please test llvm
@swift-ci please smoke test

patrykstefanski commented on clang/lib/Sema/DynamicCountPointerAssignmentAnalysis.cpp:
```diff
@@ -3485,6 +3485,9 @@ namespace clang {
   AnalysisDeclContext AC(/* AnalysisDeclContextManager */ nullptr, dcl);
+  if (isa<TemplateDecl>(dcl)) // delay processing until template has been
+                              // instantiated
+    return;
```
How does it work? Will `DynamicCountPointerAssignmentAnalysis::run` be called again when `dcl` is instantiated?

hnrklssn replied:
Yes, a new function decl will be created for the instantiated version and all the normal analyses will be run over it

Xazax-hun replied:
Is this sufficient? We could have dependent expressions when the declaration itself is not templated:
* Non-templated method of a templated struct referencing the template argument
* Lambda inside a templated function referencing the template argument
* Method of a non-templated struct nested inside a templated struct referencing the outer struct's template argument

hnrklssn replied:
Changed to an `isTemplated` call instead

ziqingluo-90 replied:
Dumb question: what's the difference between using `isTemplated` and `isDependentContext()` to rule out templates?

hnrklssn replied:
`isTemplated` covers some extra cases:
```
bool Decl::isTemplated() const {
  // A declaration is templated if it is a template or a template pattern, or
  // is within (lexcially for a friend or local function declaration,
  // semantically otherwise) a dependent context.
  if (auto *AsDC = dyn_cast<DeclContext>(this))
    return AsDC->isDependentContext();
  auto *DC = getFriendObjectKind() || isLocalExternDecl()
      ? getLexicalDeclContext() : getDeclContext();
  return DC->isDependentContext() || isTemplateDecl() ||
         getDescribedTemplateParams();
}
```

patrykstefanski commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6141,6 +6141,66 @@ static bool CheckArgLifetimeAndScope(
+    return true;
+  }
+
+  bool hadError() { return HadError; }
```
Nit: `hadError()` can be `const`.

hnrklssn replied:
Removed while transitioning to `StmtVisitor`

patrykstefanski commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6635,11 +6712,54 @@ void Sema::applyPtrCountedByEndedByAttr(Decl *D, unsigned Level,
+    // processing until the point of instantiation.
+    auto LifetimeCheck = Info.IsFPtr ? Sema::LifetimeCheckKind::None
+                                     : Sema::getLifetimeCheckKind(Var);
+    bool ScopeCheck = (Var && Var->isLocalVarDecl()) || Info.IsFPtr;
```
`LifetimeCheck` and `ScopeCheck` seems to be defined in the same way here and in `applyPtrCountedByEndedByAttr`. Maybe we can move this logic to `DynamicBoundsAttrInfo Info`?

patrykstefanski commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -8819,7 +8819,8 @@ ExprResult Sema::BuildResolvedCallExpr(Expr *Fn, NamedDecl *NDecl,
   /* TO_UPSTREAM(BoundsSafety) ON*/
-  if (getLangOpts().BoundsSafetyAttributes && FDecl) {
+  if (getLangOpts().BoundsSafetyAttributes && FDecl &&
+      !CurContext->isDependentContext()) {
```
Do `clang/test/SemaCXX/warn-unsafe-buffer-usage*` tests work with this change?

hnrklssn replied:
Yup, however they don't contain any bounds safety attributes inside templated contexts so they wouldn't be affected anyways. Adding a test case to cover this.

patrykstefanski commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6141,6 +6141,66 @@ static bool CheckArgLifetimeAndScope(
+}
+
+class EarlyLifetimeAndScopeCheck
+    : public RecursiveASTVisitor<EarlyLifetimeAndScopeCheck> {
```
Do we need `RecursiveASTVisitor`? Wouldn't `StmtVisitor` be enough?

hnrklssn replied:
IIUC with `StmtVisitor` you need to manually recurse, whereas with `RecursiveASTVisitor` traversal is already implemented so you only need to implement the non-trivial visitors. Is there a good way to do this with `StmtVisitor` without having to write a visitor for every possible type of Expr type that could occur?

patrykstefanski replied:
IIRC you can implement `VisitStmt(Stmt *S)` and call `Visit(Child)` for each `S->children()`.

hnrklssn replied:
Ah right, that seems like an improvement indeed

patrykstefanski commented:
CC @ziqingluo-90 since we were discussing templates recently.

hnrklssn committed:
address PR comments

hnrklssn commented:
@swift-ci please test

hnrklssn commented:
@swift-ci please test llvm

patrykstefanski commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6141,6 +6142,67 @@ static bool CheckArgLifetimeAndScope(
+}
+
+class EarlyLifetimeAndScopeCheck
+    : public StmtVisitor<EarlyLifetimeAndScopeCheck, bool> {
```
Even better if you use `ConstStmtVisitor`. Sorry for not mentioning this in the first review.

patrykstefanski commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6141,6 +6142,67 @@ static bool CheckArgLifetimeAndScope(
+  Sema::LifetimeCheckKind LifetimeCheck;
+  BoundsAttributedType::BoundsAttrKind AttrKind;
+  bool IsArrayType;
+  llvm::SmallPtrSet<Decl *, 16> Visited;
```
Nit: The key could be `ValueDecl *`, or `const ValueDecl *` (if you decide to switch to `ConstStmtVisitor`).

patrykstefanski commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6141,6 +6142,67 @@ static bool CheckArgLifetimeAndScope(
+
+  bool VisitStmt(Stmt *S) {
+    bool HadError = false;
+    for (auto Child : S->children())
```
Nit: I'm getting `readability-qualified-auto` warning in my editor. `auto` -> `auto *`/`const auto *`.

patrykstefanski commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -644,6 +644,43 @@ static void instantiateDependentSYCLKernelAttr(
+    Level = A->getNestedLevel();
+    ArgExpr = A->getCount();
+    Spelling = A->getSpelling();
+  } else if (auto A = dyn_cast<PtrEndedByAttr>(AL)) {
```
Nit: Same here in those autos, `auto` -> `const auto *`.

patrykstefanski commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -644,6 +644,43 @@ static void instantiateDependentSYCLKernelAttr(
+  } else
+    llvm_unreachable("unexpected late instantiated bounds safety attribute");
+
+  ExprResult InstantiatedArgExpr = S.SubstExpr(ArgExpr, TemplateArgs);
```
Is it fine not to check the result here?

Xazax-hun commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6598,48 +6683,88 @@ void Sema::applyPtrCountedByEndedByAttr(Decl *D, unsigned Level,
   if (!S.checkUInt32Argument(AL, AL.getArgAsExpr(1), Level))
     return;
+
+  if (D->getDescribedTemplate() || S.CurContext->isDependentContext()) {
```
Same question here, wondering if this is sufficient for non-templated entities nested in templates.

hnrklssn replied:
Also replaced with a call to `isTemplated`

Xazax-hun reviewed (commented):
I am not sure non-templated entities nested in templated entities are correctly handled, the very least we should have some tests for that. Added some examples of such cases inline.

hnrklssn committed:
address PR comments

hnrklssn committed:
test various nested templated contexts and fix bugs

hnrklssn commented:
@swift-ci please test llvm

hnrklssn commented:
@swift-ci please smoke test

ziqingluo-90 commented:
Just curious:  will you allow `T` being replaced with an annotated type? E.g., 
```
template<typename T> void f(T x);

void g(int * __counted_by(10) p) {
  f(p);
}
```

hnrklssn committed:
[BoundsSafety] Peek through BoundsCheckExpr in single->bidi->counted

This analysis is now done after inserting BoundsCheckExpr, so we need to
peek through them to emit these warnings where we previously only had a
BoundsSafetyPointerCast.

hnrklssn commented:
> template<typename T> void f(T x);
> 
> void g(int * __counted_by(10) p) {
>   f(p);
> }

This is currently ignored. I think it's related to how these attributes are completely ignored in casts, even C-style casts. I don't have any plans to fix this right now, but I do think that it would be good to get there eventually.

hnrklssn commented:
@swift-ci please test llvm

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please test

hnrklssn commented:
windows failure seems unrelated

hnrklssn commented:
@swift-ci please test llvm
